pinot-plugins/pinot-file-system/pinot-s3/src/main/java/org/apache/pinot/plugin/filesystem/S3PinotFS.java

```diff
@@ -0,0 +1,485 @@
+        try {
+            String encodedUrl = null;
+            try {
+                encodedUrl = URLEncoder.encode(srcUri.getHost() + srcUri.getPath(), StandardCharsets.UTF_8.toString());
+            } catch (UnsupportedEncodingException e) {
+                LOGGER.info("URL could not be encoded: {}", e.getMessage());
+            }
+
+            String dstPath = sanitizePath(dstUri.getPath());
+            CopyObjectRequest copyReq = CopyObjectRequest.builder()
+                    .copySource(encodedUrl)
+                    .destinationBucket(dstUri.getHost())
+                    .destinationKey(dstPath)
+                    .build();
+
+            CopyObjectResponse copyObjectResponse = s3Client.copyObject(copyReq);
+            return copyObjectResponse.sdkHttpResponse().isSuccessful();
+        }catch(S3Exception e){
+            throw new IOException(e);
+        }
+    }
+
+    @Override
+    public boolean mkdir(URI uri) throws IOException {
+        LOGGER.info("mkdir {}", uri);
+        try {
+            requireNonNull(uri, "uri is null");
+            String path = normalizeToDirectoryPrefix(uri);
+            // Bucket root directory already exists and cannot be created
+            if (path.equals(DELIMITER)) {
+                return true;
+            }
+
+            PutObjectRequest putObjectRequest = PutObjectRequest.builder()
+                    .bucket(uri.getHost())
+                    .key(path)
+                    .build();
+
+            PutObjectResponse putObjectResponse = s3Client.putObject(putObjectRequest, RequestBody.fromBytes(new byte[0]));
+
+            return putObjectResponse.sdkHttpResponse().isSuccessful();
+        } catch (Throwable t) {
+            throw new IOException(t);
+        }
+    }
+
+    @Override
+    public boolean delete(URI segmentUri, boolean forceDelete) throws IOException {
+        LOGGER.info("Deleting uri {} force {}", segmentUri, forceDelete);
+        try {
+            if (isDirectory(segmentUri)) {
+                if (!forceDelete) {
+                    checkState(isEmptyDirectory(segmentUri), "ForceDelete flag is not set and directory '%s' is not empty", segmentUri);
+                }
+                String prefix = normalizeToDirectoryPrefix(segmentUri);
+                ListObjectsV2Response listObjectsV2Response;
+                ListObjectsV2Request.Builder listObjectsV2RequestBuilder = ListObjectsV2Request.builder()
+                        .bucket(segmentUri.getHost());
+
+                if (prefix.equals(DELIMITER)) {
+                    ListObjectsV2Request listObjectsV2Request = listObjectsV2RequestBuilder.build();
+                    listObjectsV2Response = s3Client.listObjectsV2(listObjectsV2Request);
+                } else {
+                    ListObjectsV2Request listObjectsV2Request = listObjectsV2RequestBuilder.prefix(prefix).build();
+                    listObjectsV2Response = s3Client.listObjectsV2(listObjectsV2Request);
```

Comment:
nit: `listObjectsV2Response = s3Client.listObjectsV2(listObjectsV2Request);` seems repetitive in both blocks. Move it below if-else.

Reply:
done